### ISSUE TITLE
no-unused-vars: Count JSDoc comments and type annotations as using variables

### ISSUE THREAD

**Issue by AnthonyWeston** [eslint/eslint#12541]
**What rule do you want to change?**
no-unused-vars

**Does this change cause the rule to produce more or fewer warnings?**
Fewer warnings - It will also reduce the need to add comments to disable/re-enable eslint rules when importing types, etc. from other files for use in JSDoc coments.

**How will the change be implemented? (New option, new default behavior, etc.)?**
I would personally prefer default behavior, but if that breaks things an option would be fine too.

**Please provide some example code that this change will affect:**

```
import { Foo, Bar, Baz, BazBuilder } from 'foo'

 /**
  * Creates a Bar from a Foo
  * @param { Foo } foo 
  * @returns { Bar } bar
  */
export function useFoo(foo) {
    /** @type { Baz } */
    const baz = BazBuilder.build()

    return foo.createBar(baz)
}
```

**What does the rule currently do for this code?**
It flags the Foo, Bar, and Baz imports as being unused, even though 
appear in the JSdoc comment and type annotation.

**What will the rule do after it's changed?**
It won't flag the Foo, Bar, and Baz imports as being unsued because they are in the JSDoc comment and type annotation.

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by kaicataldo** [comment]
We have [deprecated JSDoc-aware features](https://eslint.org/blog/2018/11/jsdoc-end-of-life) in ESLint core. This definitely seems useful, though! I wonder if making an issue for [`eslint-plugin-jsdoc`](https://github.com/gajus/eslint-plugin-jsdoc) might be the right course of action here.

**Comment by golopot** [comment]
I recommend importing JSDoc types using this syntax: 
```js
/** @type {import('foo').Baz} */
const baz = BazBuilder.build()
```
